solver/src/solver/naive_solver/multi_order_solver.rs

```diff
@@ -115,33 +116,62 @@ fn solve_without_uniswap(
 /// The clearing price is the effective exchange rate used by the AMM interaction.
 ///
 fn solve_with_uniswap(
-    orders: impl Iterator<Item = LimitOrder> + Clone,
+    orders: &[LimitOrder],
     pool: &ConstantProductOrder,
     shortage: &TokenContext,
     excess: &TokenContext,
 ) -> Option<Settlement> {
+    let uniswap_in_token = excess.address;
+    let uniswap_out_token = shortage.address;
+
     let uniswap_out = compute_uniswap_out(shortage, excess, pool.fee)?;
     let uniswap_in = compute_uniswap_in(uniswap_out.clone(), shortage, excess, pool.fee);
 
     let uniswap_out = big_rational_to_u256(&uniswap_out).ok()?;
     let uniswap_in = big_rational_to_u256(&uniswap_in).ok()?;
 
     let mut settlement = Settlement::new(maplit::hashmap! {
-        shortage.address => uniswap_in,
-        excess.address => uniswap_out,
+        uniswap_in_token => uniswap_out,
+        uniswap_out_token => uniswap_in,
     });
     for order in orders {
         settlement
-            .with_liquidity(&order, order.full_execution_amount())
+            .with_liquidity(order, order.full_execution_amount())
             .ok()?;
     }
 
+    // Because the smart contracts round in the favour of the traders, it could
+    // be that we actually require a bit more from the Uniswap pool in order to
+    // pay out all proceeds. We move the rounding error to the sell token so that
+    // it either comes out of the fees or existing buffers. Compute that amount:
+    let uniswap_out_with_rounding = big_int_to_u256(&settlement.executed_trades().fold(
```

Comment:
In the description you mention the extra amount is capped at 1 wei per trade. So would an alternative implementation here simply add +- 1 to uniswap_out per trade here? Just trying to see if I understand this right.

Reply:
It would be to add `orders.len()` to the final output amount. Its not always needed though (if there is one order for example, then there won't be any rounding issues).